code/modules/power/batteryrack.dm

```diff
@@ -1,234 +1,313 @@
+		var/celldiff
+		// Take amount of power to transfer from the cell with smaller maxcharge
+		if(most.maxcharge > least.maxcharge)
+			celldiff = (least.maxcharge / 100) * percentdiff
 		else
-			user.set_machine(src)
-			interact(user)
-			return 1
-	return
+			celldiff = (most.maxcharge / 100) * percentdiff
+		celldiff = between(0, celldiff, max_transfer_rate * CELLRATE)
+		// Ensure we don't transfer more energy than the most charged cell has, and that the least charged cell can input.
+		celldiff = min(min(celldiff, most.charge), least.maxcharge - least.charge)
+		least.give(most.use(celldiff))
 
+/obj/machinery/power/smes/batteryrack/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	var/data[0]
 
-//The shitty one that will blow up.
-/obj/machinery/power/smes/batteryrack/makeshift
-	name = "makeshift PSU"
-	desc = "A rack of batteries connected by a mess of wires posing as a PSU."
-	circuit = /obj/item/weapon/circuitboard/ghettosmes
-	var/overcharge_percent = 0
+	data["mode"] = mode
+	data["transfer_max"] = max_transfer_rate
+	data["output_load"] = round(output_used)
+	data["input_load"] = round(input_available)
+	data["equalise"] = equalise
+	data["blink_tick"] = ui_tick
+	data["cells_max"] = max_cells
+	data["cells_cur"] = internal_cells.len
+	var/list/cells = list()
+	var/cell_index = 0
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		var/list/cell[0]
+		cell["slot"] = cell_index + 1
+		cell["used"] = 1
+		cell["percentage"] = round(C.percent(), 0.01)
+		cell["id"] = C.c_uid
+		cell_index++
+		cells += list(cell)
+	while(cell_index < PSU_MAXCELLS)
+		var/list/cell[0]
+		cell["slot"] = cell_index + 1
+		cell["used"] = 0
+		cell_index++
+		cells += list(cell)
+	data["cells_list"] = cells
 
+	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "psu.tmpl", "Cell Rack PSU", 500, 430)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
 
-/obj/machinery/power/smes/batteryrack/makeshift/add_parts()
-	component_parts = list()
-	component_parts += new /obj/item/weapon/cell/high
-	component_parts += new /obj/item/weapon/cell/high
-	component_parts += new /obj/item/weapon/cell/high
-	return
+/obj/machinery/power/smes/batteryrack/dismantle()
+	for(var/obj/item/weapon/cell/C in internal_cells)
+		C.forceMove(get_turf(src))
+		internal_cells -= C
+	return ..()
 
+/obj/machinery/power/smes/batteryrack/attackby(var/obj/item/weapon/W as obj, var/mob/user as mob)
+	if(!..())
+		return 0
+	if(default_deconstruction_crowbar(user, W))
+		return
+	if(default_part_replacement(user, W))
+		return
+	if(istype(W, /obj/item/weapon/cell)) // ID Card, try to insert it.
+		if(insert_cell(W, user))
+			to_chat(user, "You insert \the [W] into \the [src].")
```

Comment:
filter tags in here, please wrap the message in `<span class='filter_[category]'> [...]</span>`